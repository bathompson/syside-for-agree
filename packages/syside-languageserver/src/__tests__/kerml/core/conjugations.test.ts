/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

const Common = `
class Original {
    in feature Input;
}
class Conjugate1;
class Conjugate2;
`;

test("conjugation can be parsed", async () => {
    return expect(
        Common +
            `
    conjugation c1 conjugate Conjugate1 conjugates Original;
    conjugation c2 conjugate Conjugate2 ~ Original {
        doc /* same as c1 */
    }`
    ).toParseKerML("snapshot");
});

test.each(["conjugates", "~"])(
    "conjugation may be omitted without identifiers with '%s'",
    async (token: string) => {
        return expect(Common + `conjugate Conjugate1 ${token} Original;`).toParseKerML("snapshot");
    }
);

test.each(["conjugates", "~"])(
    "type can declare owned conjugations with '%s'",
    async (token: string) => {
        return expect(
            `class Original;
    class Conjugate1 ${token} Original;`
        ).toParseKerML("snapshot");
    }
);
