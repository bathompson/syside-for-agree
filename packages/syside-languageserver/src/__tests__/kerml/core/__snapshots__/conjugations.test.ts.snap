// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conjugation can be parsed: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [
            {
              "$type": "FeatureMembership",
              "element": {
                "$type": "Feature",
                "children": [],
                "declaredName": "Input",
                "direction": "in",
                "heritage": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "prefixes": [],
                "typeRelationships": [],
              },
              "elements": [],
            },
          ],
          "declaredName": "Original",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate1",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate2",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Conjugation",
          "declaredName": "c1",
          "elements": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [
                    {
                      "$type": "FeatureMembership",
                      "element": {
                        "$type": "Feature",
                        "children": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "heritage": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "prefixes": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                    },
                  ],
                  "declaredName": "Original",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [],
                  "declaredName": "Conjugate1",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Conjugation",
          "declaredName": "c2",
          "elements": [
            {
              "$type": "Annotation",
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* same as c1 */",
              },
              "elements": [],
            },
          ],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [
                    {
                      "$type": "FeatureMembership",
                      "element": {
                        "$type": "Feature",
                        "children": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "heritage": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "prefixes": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                    },
                  ],
                  "declaredName": "Original",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [],
                  "declaredName": "Conjugate2",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;

exports[`conjugation may be omitted without identifiers with '~': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [
            {
              "$type": "FeatureMembership",
              "element": {
                "$type": "Feature",
                "children": [],
                "declaredName": "Input",
                "direction": "in",
                "heritage": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "prefixes": [],
                "typeRelationships": [],
              },
              "elements": [],
            },
          ],
          "declaredName": "Original",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate1",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate2",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Conjugation",
          "elements": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [
                    {
                      "$type": "FeatureMembership",
                      "element": {
                        "$type": "Feature",
                        "children": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "heritage": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "prefixes": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                    },
                  ],
                  "declaredName": "Original",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [],
                  "declaredName": "Conjugate1",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;

exports[`conjugation may be omitted without identifiers with 'conjugates': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [
            {
              "$type": "FeatureMembership",
              "element": {
                "$type": "Feature",
                "children": [],
                "declaredName": "Input",
                "direction": "in",
                "heritage": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "prefixes": [],
                "typeRelationships": [],
              },
              "elements": [],
            },
          ],
          "declaredName": "Original",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate1",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate2",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Conjugation",
          "elements": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [
                    {
                      "$type": "FeatureMembership",
                      "element": {
                        "$type": "Feature",
                        "children": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "heritage": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "prefixes": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                    },
                  ],
                  "declaredName": "Original",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "children": [],
                  "declaredName": "Conjugate1",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;

exports[`type can declare owned conjugations with '~': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Original",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate1",
          "heritage": [
            {
              "$type": "Conjugation",
              "elements": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "children": [],
                      "declaredName": "Original",
                      "heritage": [],
                      "isSufficient": false,
                      "prefixes": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;

exports[`type can declare owned conjugations with 'conjugates': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Original",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Class",
          "children": [],
          "declaredName": "Conjugate1",
          "heritage": [
            {
              "$type": "Conjugation",
              "elements": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "children": [],
                      "declaredName": "Original",
                      "heritage": [],
                      "isSufficient": false,
                      "prefixes": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;
